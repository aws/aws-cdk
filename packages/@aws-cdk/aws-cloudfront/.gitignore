*.js
*.d.ts
tsconfig.json
*.js.map
*.generated.ts
dist
coverage
.nyc_output
.jsii

.LAST_BUILD
nyc.config.js
.LAST_PACKAGE
*.snk
!.eslintrc.js
!jest.config.js

!lib/experimental/edge-function/index.js

junit.xml
