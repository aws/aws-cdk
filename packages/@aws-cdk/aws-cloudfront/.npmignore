# Don't include original .ts files when doing `npm pack`
*.ts
!*.d.ts
cfn
dist
coverage
.nyc_output
*.tgz

.LAST_PACKAGE
.LAST_BUILD
!*.js

# Include .jsii
!.jsii

*.snk

*.tsbuildinfo

tsconfig.json
.eslintrc.js

# exclude cdk artifacts
**/cdk.out
jest.config.js

junit.xml
test/