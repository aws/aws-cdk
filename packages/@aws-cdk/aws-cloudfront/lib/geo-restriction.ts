/**
 * Controls the countries in which content is distributed.
 */
export class GeoRestriction {

  /**
   * Whitelist specific countries which you want CloudFront to distribute your content.
   *
   * @param locations Two-letter, uppercase country code for a country
   * that you want to whitelist. Include one element for each country.
   * See ISO 3166-1-alpha-2 code on the *International Organization for Standardization* website
   */
  public static whitelist(...locations: string[]) {
    return new GeoRestriction('whitelist', GeoRestriction.validateLocations(locations));
  }

  /**
   * Blacklist specific countries which you don't want CloudFront to distribute your content.
   *
   * @param locations Two-letter, uppercase country code for a country
   * that you want to blacklist. Include one element for each country.
   * See ISO 3166-1-alpha-2 code on the *International Organization for Standardization* website
   */
  public static blacklist(...locations: string[]) {
    return new GeoRestriction('blacklist', GeoRestriction.validateLocations(locations));
  }

  private static LOCATION_REGEX = /^[A-Z]{2}$/;

  private static validateLocations(locations: string[]) {
    if (locations.length === 0) {
      throw new Error('Should provide at least 1 location');
    }
    locations.forEach(location => {
      if (!GeoRestriction.LOCATION_REGEX.test(location)) {
        // eslint-disable-next-line max-len
        throw new Error(`Invalid location format for location: ${location}, location should be two-letter and uppercase country ISO 3166-1-alpha-2 code`);
      }
    });
    return locations;
  }

  /**
   * Creates an instance of GeoRestriction for internal use
   *
   * @param restrictionType Specifies the restriction type to impose (whitelist or blacklist)
   * @param locations Two-letter, uppercase country code for a country
   * that you want to whitelist/blacklist. Include one element for each country.
   * See ISO 3166-1-alpha-2 code on the *International Organization for Standardization* website
   */
  private constructor(readonly restrictionType: 'whitelist' | 'blacklist', readonly locations: string[]) {}
}
