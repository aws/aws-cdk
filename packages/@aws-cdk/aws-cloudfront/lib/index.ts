export * from './cache-policy';
export * from './distribution';
export * from './geo-restriction';
export * from './origin';
export * from './origin-access-identity';
export * from './origin-request-policy';
export * from './web-distribution';

export * as experimental from './experimental';

// AWS::CloudFront CloudFormation Resources:
export * from './cloudfront.generated';
