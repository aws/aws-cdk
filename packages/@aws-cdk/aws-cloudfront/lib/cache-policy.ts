import { Duration, Names, Resource, Stack, Token } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnCachePolicy } from './cloudfront.generated';

/**
 * Represents a Cache Policy
 */
export interface ICachePolicy {
  /**
   * The ID of the cache policy
   * @attribute
   */
  readonly cachePolicyId: string;
}

/**
 * Properties for creating a Cache Policy
 */
export interface CachePolicyProps {
  /**
   * A unique name to identify the cache policy.
   * The name must only include '-', '_', or alphanumeric characters.
   * @default - generated from the `id`
   */
  readonly cachePolicyName?: string;

  /**
   * A comment to describe the cache policy.
   * @default - no comment
   */
  readonly comment?: string;

  /**
   * The default amount of time for objects to stay in the CloudFront cache.
   * Only used when the origin does not send Cache-Control or Expires headers with the object.
   * @default - The greater of 1 day and ``minTtl``
   */
  readonly defaultTtl?: Duration;

  /**
   * The minimum amount of time for objects to stay in the CloudFront cache.
   * @default Duration.seconds(0)
   */
  readonly minTtl?: Duration;

  /**
   * The maximum amount of time for objects to stay in the CloudFront cache.
   * CloudFront uses this value only when the origin sends Cache-Control or Expires headers with the object.
   * @default - The greater of 1 year and ``defaultTtl``
   */
  readonly maxTtl?: Duration;

  /**
   * Determines whether any cookies in viewer requests are included in the cache key and automatically included in requests that CloudFront sends to the origin.
   * @default CacheCookieBehavior.none()
   */
  readonly cookieBehavior?: CacheCookieBehavior;

  /**
   * Determines whether any HTTP headers are included in the cache key and automatically included in requests that CloudFront sends to the origin.
   * @default CacheHeaderBehavior.none()
   */
  readonly headerBehavior?: CacheHeaderBehavior;

  /**
   * Determines whether any query strings are included in the cache key and automatically included in requests that CloudFront sends to the origin.
   * @default CacheQueryStringBehavior.none()
   */
  readonly queryStringBehavior?: CacheQueryStringBehavior;

  /**
   * Whether to normalize and include the `Accept-Encoding` header in the cache key when the `Accept-Encoding` header is 'gzip'.
   * @default false
   */
  readonly enableAcceptEncodingGzip?: boolean;

  /**
   * Whether to normalize and include the `Accept-Encoding` header in the cache key when the `Accept-Encoding` header is 'br'.
   * @default false
   */
  readonly enableAcceptEncodingBrotli?: boolean;
}

/**
 * A Cache Policy configuration.
 *
 * @resource AWS::CloudFront::CachePolicy
 */
export class CachePolicy extends Resource implements ICachePolicy {

  /**
   * Optimize cache efficiency by minimizing the values that CloudFront includes in the cache key.
   * Query strings and cookies are not included in the cache key, and only the normalized 'Accept-Encoding' header is included.
   */
  public static readonly CACHING_OPTIMIZED = CachePolicy.fromManagedCachePolicy('658327ea-f89d-4fab-a63d-7e88639e58f6');
  /**
   * Optimize cache efficiency by minimizing the values that CloudFront includes in the cache key.
   * Query strings and cookies are not included in the cache key, and only the normalized 'Accept-Encoding' header is included.
   * Disables cache compression.
   */
  public static readonly CACHING_OPTIMIZED_FOR_UNCOMPRESSED_OBJECTS = CachePolicy.fromManagedCachePolicy('b2884449-e4de-46a7-ac36-70bc7f1ddd6d');
  /** Disables caching. This policy is useful for dynamic content and for requests that are not cacheable. */
  public static readonly CACHING_DISABLED = CachePolicy.fromManagedCachePolicy('4135ea2d-6df8-44a3-9df3-4b5a84be39ad');
  /** Designed for use with an origin that is an AWS Elemental MediaPackage endpoint. */
  public static readonly ELEMENTAL_MEDIA_PACKAGE = CachePolicy.fromManagedCachePolicy('08627262-05a9-4f76-9ded-b50ca2e3a84f');

  /** Imports a Cache Policy from its id. */
  public static fromCachePolicyId(scope: Construct, id: string, cachePolicyId: string): ICachePolicy {
    return new class extends Resource implements ICachePolicy {
      public readonly cachePolicyId = cachePolicyId;
    }(scope, id);
  }

  /** Use an existing managed cache policy. */
  private static fromManagedCachePolicy(managedCachePolicyId: string): ICachePolicy {
    return new class implements ICachePolicy {
      public readonly cachePolicyId = managedCachePolicyId;
    }();
  }

  public readonly cachePolicyId: string;

  constructor(scope: Construct, id: string, props: CachePolicyProps = {}) {
    super(scope, id, {
      physicalName: props.cachePolicyName,
    });

    const cachePolicyName = props.cachePolicyName ?? `${Names.uniqueId(this)}-${Stack.of(this).region}`;
    if (!Token.isUnresolved(cachePolicyName) && !cachePolicyName.match(/^[\w-]+$/i)) {
      throw new Error(`'cachePolicyName' can only include '-', '_', and alphanumeric characters, got: '${props.cachePolicyName}'`);
    }

    const minTtl = (props.minTtl ?? Duration.seconds(0)).toSeconds();
    const defaultTtl = Math.max((props.defaultTtl ?? Duration.days(1)).toSeconds(), minTtl);
    const maxTtl = Math.max((props.maxTtl ?? Duration.days(365)).toSeconds(), defaultTtl);

    const resource = new CfnCachePolicy(this, 'Resource', {
      cachePolicyConfig: {
        name: cachePolicyName,
        comment: props.comment,
        minTtl,
        maxTtl,
        defaultTtl,
        parametersInCacheKeyAndForwardedToOrigin: this.renderCacheKey(props),
      },
    });

    this.cachePolicyId = resource.ref;
  }

  private renderCacheKey(props: CachePolicyProps): CfnCachePolicy.ParametersInCacheKeyAndForwardedToOriginProperty {
    const cookies = props.cookieBehavior ?? CacheCookieBehavior.none();
    const headers = props.headerBehavior ?? CacheHeaderBehavior.none();
    const queryStrings = props.queryStringBehavior ?? CacheQueryStringBehavior.none();

    return {
      cookiesConfig: {
        cookieBehavior: cookies.behavior,
        cookies: cookies.cookies,
      },
      headersConfig: {
        headerBehavior: headers.behavior,
        headers: headers.headers,
      },
      enableAcceptEncodingGzip: props.enableAcceptEncodingGzip ?? false,
      enableAcceptEncodingBrotli: props.enableAcceptEncodingBrotli ?? false,
      queryStringsConfig: {
        queryStringBehavior: queryStrings.behavior,
        queryStrings: queryStrings.queryStrings,
      },
    };
  }
}

/**
 * Determines whether any cookies in viewer requests are included in the cache key and
 * automatically included in requests that CloudFront sends to the origin.
 */
export class CacheCookieBehavior {
  /**
   * Cookies in viewer requests are not included in the cache key and
   * are not automatically included in requests that CloudFront sends to the origin.
   */
  public static none() { return new CacheCookieBehavior('none'); }

  /**
   * All cookies in viewer requests are included in the cache key and are automatically included in requests that CloudFront sends to the origin.
   */
  public static all() { return new CacheCookieBehavior('all'); }

  /**
   * Only the provided `cookies` are included in the cache key and automatically included in requests that CloudFront sends to the origin.
   */
  public static allowList(...cookies: string[]) {
    if (cookies.length === 0) {
      throw new Error('At least one cookie to allow must be provided');
    }
    return new CacheCookieBehavior('whitelist', cookies);
  }

  /**
   * All cookies except the provided `cookies` are included in the cache key and
   * automatically included in requests that CloudFront sends to the origin.
   */
  public static denyList(...cookies: string[]) {
    if (cookies.length === 0) {
      throw new Error('At least one cookie to deny must be provided');
    }
    return new CacheCookieBehavior('allExcept', cookies);
  }

  /** The behavior of cookies: allow all, none, an allow list, or a deny list. */
  public readonly behavior: string;
  /** The cookies to allow or deny, if the behavior is an allow or deny list. */
  public readonly cookies?: string[];

  private constructor(behavior: string, cookies?: string[]) {
    this.behavior = behavior;
    this.cookies = cookies;
  }
}

/**
 * Determines whether any HTTP headers are included in the cache key and automatically included in requests that CloudFront sends to the origin.
 */
export class CacheHeaderBehavior {
  /** HTTP headers are not included in the cache key and are not automatically included in requests that CloudFront sends to the origin. */
  public static none() { return new CacheHeaderBehavior('none'); }
  /** Listed headers are included in the cache key and are automatically included in requests that CloudFront sends to the origin. */
  public static allowList(...headers: string[]) {
    if (headers.length === 0) {
      throw new Error('At least one header to allow must be provided');
    }
    if (headers.length > 10) {
      throw new Error(`Maximum allowed headers in Cache Policy is 10; got ${headers.length}.`);
    }
    return new CacheHeaderBehavior('whitelist', headers);
  }

  /** If the no headers will be passed, or an allow list of headers. */
  public readonly behavior: string;
  /** The headers for the allow/deny list, if applicable. */
  public readonly headers?: string[];

  private constructor(behavior: string, headers?: string[]) {
    this.behavior = behavior;
    this.headers = headers;
  }
}

/**
 * Determines whether any URL query strings in viewer requests are included in the cache key
 * and automatically included in requests that CloudFront sends to the origin.
 */
export class CacheQueryStringBehavior {
  /**
   * Query strings in viewer requests are not included in the cache key and
   * are not automatically included in requests that CloudFront sends to the origin.
   */
  public static none() { return new CacheQueryStringBehavior('none'); }

  /**
   * All query strings in viewer requests are included in the cache key and are automatically included in requests that CloudFront sends to the origin.
   */
  public static all() { return new CacheQueryStringBehavior('all'); }

  /**
   * Only the provided `queryStrings` are included in the cache key and automatically included in requests that CloudFront sends to the origin.
   */
  public static allowList(...queryStrings: string[]) {
    if (queryStrings.length === 0) {
      throw new Error('At least one query string to allow must be provided');
    }
    return new CacheQueryStringBehavior('whitelist', queryStrings);
  }

  /**
   * All query strings except the provided `queryStrings` are included in the cache key and
   * automatically included in requests that CloudFront sends to the origin.
   */
  public static denyList(...queryStrings: string[]) {
    if (queryStrings.length === 0) {
      throw new Error('At least one query string to deny must be provided');
    }
    return new CacheQueryStringBehavior('allExcept', queryStrings);
  }

  /** The behavior of query strings -- allow all, none, only an allow list, or a deny list. */
  public readonly behavior: string;
  /** The query strings to allow or deny, if the behavior is an allow or deny list. */
  public readonly queryStrings?: string[];

  private constructor(behavior: string, queryStrings?: string[]) {
    this.behavior = behavior;
    this.queryStrings = queryStrings;
  }
}
