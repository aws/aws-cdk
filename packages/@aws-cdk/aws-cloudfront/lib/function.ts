import * as fs from 'fs';
import { IResource, Names, Resource, Stack } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnFunction } from './cloudfront.generated';

/**
 * Represents the function's source code
 */
export abstract class FunctionCode {

  /**
   * Inline code for function
   * @returns code object with inline code.
   * @param code The actual function code
   */
  public static fromInline(code: string): FunctionCode {
    return new InlineCode(code);
  }

  /**
   * Code from external file for function
   * @returns code object with contents from file.
   * @param options the options for the external file
   */
  public static fromFile(options: FileCodeOptions): FunctionCode {
    return new FileCode(options);
  }

  /**
   * renders the function code
   */
  public abstract render(): string;
}

/**
 * Options when reading the function's code from an external file
 */
export interface FileCodeOptions {
  /**
   * The path of the file to read the code from
   */
  readonly filePath: string;
}

/**
 * Represents the function's source code as inline code
 */
class InlineCode extends FunctionCode {

  constructor(private code: string) {
    super();
  }

  public render(): string {
    return this.code;
  }
}


/**
 * Represents the function's source code loaded from an external file
 */
class FileCode extends FunctionCode {

  constructor(private options: FileCodeOptions) {
    super();
  }

  public render(): string {
    return fs.readFileSync(this.options.filePath, { encoding: 'utf-8' });
  }
}

/**
 * Represents a CloudFront Function
 */
export interface IFunction extends IResource {
  /**
   * The name of the function.
   * @attribute
   */
  readonly functionName: string;

  /**
   * The ARN of the function.
   * @attribute
   */
  readonly functionArn: string;
}

/**
 * Attributes of an existing CloudFront Function to import it
 */
export interface FunctionAttributes {
  /**
   * The name of the function.
   */
  readonly functionName: string;

  /**
   * The ARN of the function.
   */
  readonly functionArn: string;
}

/**
 * Properties for creating a CloudFront Function
 */
export interface FunctionProps {
  /**
   * A name to identify the function.
   * @default - generated from the `id`
   */
  readonly functionName?: string;

  /**
   * A comment to describe the function.
   * @default - same as `functionName`
   */
  readonly comment?: string;

  /**
   * The source code of the function.
   */
  readonly code: FunctionCode;
}

/**
 * A CloudFront Function
 *
 * @resource AWS::CloudFront::Function
 */
export class Function extends Resource implements IFunction {

  /** Imports a function by its name and ARN */
  public static fromFunctionAttributes(scope: Construct, id: string, attrs: FunctionAttributes): IFunction {
    return new class extends Resource implements IFunction {
      public readonly functionName = attrs.functionName;
      public readonly functionArn = attrs.functionArn;
    }(scope, id);
  }

  /**
   * the name of the CloudFront function
   * @attribute
   */
  public readonly functionName: string;
  /**
   * the ARN of the CloudFront function
   * @attribute
   */
  public readonly functionArn: string;
  /**
   * the deployment stage of the CloudFront function
   * @attribute
   */
  public readonly functionStage: string;

  constructor(scope: Construct, id: string, props: FunctionProps) {
    super(scope, id);

    this.functionName = props.functionName ?? this.generateName();

    const resource = new CfnFunction(this, 'Resource', {
      autoPublish: true,
      functionCode: props.code.render(),
      functionConfig: {
        comment: props.comment ?? this.functionName,
        runtime: 'cloudfront-js-1.0',
      },
      name: this.functionName,
    });

    this.functionArn = resource.attrFunctionArn;
    this.functionStage = resource.attrStage;
  }

  private generateName(): string {
    const name = Stack.of(this).region + Names.uniqueId(this);
    if (name.length > 64) {
      return name.substring(0, 32) + name.substring(name.length - 32);
    }
    return name;
  }
}

/**
 * The type of events that a CloudFront function can be invoked in response to.
 */
export enum FunctionEventType {

  /**
   * The viewer-request specifies the incoming request
   */
  VIEWER_REQUEST = 'viewer-request',

  /**
   * The viewer-response specifies the outgoing response
   */
  VIEWER_RESPONSE = 'viewer-response',
}

/**
 * Represents a CloudFront function and event type when using CF Functions.
 * The type of the {@link AddBehaviorOptions.functionAssociations} property.
 */
export interface FunctionAssociation {
  /**
   * The CloudFront function that will be invoked.
   */
  readonly function: IFunction;

  /** The type of event which should invoke the function. */
  readonly eventType: FunctionEventType;
}
