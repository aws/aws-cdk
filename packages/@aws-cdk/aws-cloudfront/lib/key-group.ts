import { IResource, Names, Resource } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnKeyGroup } from './cloudfront.generated';
import { IPublicKey } from './public-key';

/**
 * Represents a Key Group
 */
export interface IKeyGroup extends IResource {
  /**
   * The ID of the key group.
   * @attribute
   */
  readonly keyGroupId: string;
}

/**
 * Properties for creating a Public Key
 */
export interface KeyGroupProps {
  /**
   * A name to identify the key group.
   * @default - generated from the `id`
   */
  readonly keyGroupName?: string;

  /**
   * A comment to describe the key group.
   * @default - no comment
   */
  readonly comment?: string;

  /**
   * A list of public keys to add to the key group.
   */
  readonly items: IPublicKey[];
}

/**
 * A Key Group configuration
 *
 * @resource AWS::CloudFront::KeyGroup
 */
export class KeyGroup extends Resource implements IKeyGroup {

  /** Imports a Key Group from its id. */
  public static fromKeyGroupId(scope: Construct, id: string, keyGroupId: string): IKeyGroup {
    return new class extends Resource implements IKeyGroup {
      public readonly keyGroupId = keyGroupId;
    }(scope, id);
  }
  public readonly keyGroupId: string;

  constructor(scope: Construct, id: string, props: KeyGroupProps) {
    super(scope, id);

    const resource = new CfnKeyGroup(this, 'Resource', {
      keyGroupConfig: {
        name: props.keyGroupName ?? this.generateName(),
        comment: props.comment,
        items: props.items.map(key => key.publicKeyId),
      },
    });

    this.keyGroupId = resource.ref;
  }

  private generateName(): string {
    const name = Names.uniqueId(this);
    if (name.length > 80) {
      return name.substring(0, 40) + name.substring(name.length - 40);
    }
    return name;
  }
}
