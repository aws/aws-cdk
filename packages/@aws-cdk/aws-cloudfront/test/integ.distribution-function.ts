import * as cdk from '@aws-cdk/core';
import * as cloudfront from '../lib';
import { TestOrigin } from './test-origin';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'integ-distribution-function', { env: { region: 'eu-west-1' } });

const cfFunction = new cloudfront.Function(stack, 'Function', {
  code: cloudfront.FunctionCode.fromInline('function handler(event) { return event.request }'),
});

new cloudfront.Distribution(stack, 'Dist', {
  defaultBehavior: {
    origin: new TestOrigin('www.example.com'),
    cachePolicy: cloudfront.CachePolicy.CACHING_DISABLED,
    functionAssociations: [{
      function: cfFunction,
      eventType: cloudfront.FunctionEventType.VIEWER_REQUEST,
    }],
  },
});

new cdk.CfnOutput(stack, 'FunctionArn', { value: cfFunction.functionArn });
new cdk.CfnOutput(stack, 'FunctionStage', { value: cfFunction.functionStage });

app.synth();
