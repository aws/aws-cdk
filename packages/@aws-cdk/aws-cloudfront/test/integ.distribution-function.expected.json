{
  "Resources": {
    "Function76856677": {
      "Type": "AWS::CloudFront::Function",
      "Properties": {
        "Name": "eu-west-1integdistributionfunctionFunctionDCD62A02",
        "AutoPublish": true,
        "FunctionCode": "function handler(event) { return event.request }",
        "FunctionConfig": {
          "Comment": "eu-west-1integdistributionfunctionFunctionDCD62A02",
          "Runtime": "cloudfront-js-1.0"
        }
      }
    },
    "DistB3B78991": {
      "Type": "AWS::CloudFront::Distribution",
      "Properties": {
        "DistributionConfig": {
          "DefaultCacheBehavior": {
            "CachePolicyId": "4135ea2d-6df8-44a3-9df3-4b5a84be39ad",
            "Compress": true,
            "FunctionAssociations": [
              {
                "EventType": "viewer-request",
                "FunctionARN": {
                  "Fn::GetAtt": [
                    "Function76856677",
                    "FunctionARN"
                  ]
                }
              }
            ],
            "TargetOriginId": "integdistributionfunctionDistOrigin1D1E9DF17",
            "ViewerProtocolPolicy": "allow-all"
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "CustomOriginConfig": {
                "OriginProtocolPolicy": "https-only"
              },
              "DomainName": "www.example.com",
              "Id": "integdistributionfunctionDistOrigin1D1E9DF17"
            }
          ]
        }
      }
    }
  },
  "Outputs": {
    "FunctionArn": {
      "Value": {
        "Fn::GetAtt": [
          "Function76856677",
          "FunctionARN"
        ]
      }
    },
    "FunctionStage": {
      "Value": {
        "Fn::GetAtt": [
          "Function76856677",
          "Stage"
        ]
      }
    }
  }
}